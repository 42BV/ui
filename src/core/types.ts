/* istanbul ignore file */

export type Color =
  | 'primary'
  | 'secondary'
  | 'success'
  | 'info'
  | 'warning'
  | 'danger'
  | 'link'
  | 'muted'
  | 'dark'
  | 'light';

export type ModalSize = 'sm' | 'md' | 'lg' | 'xl';

type Side = 'top' | 'right' | 'bottom' | 'left';
type StartOrEnd = 'start' | 'end';
export type TippyPlacement = Side | `${Side}-${StartOrEnd}`;
