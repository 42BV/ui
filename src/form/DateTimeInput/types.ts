// See https://momentjs.com/docs/#/displaying/format/ for formatting options.
export type DateFormat = string | false; // When false it should not be used
export type TimeFormat = string | false; // When false it should not be used
