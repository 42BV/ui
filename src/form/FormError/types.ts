export type OnChange = (hasErrors: boolean) => void;