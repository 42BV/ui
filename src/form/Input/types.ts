export type Position = 'left' | 'right';
export type Mask = (string | RegExp)[];
