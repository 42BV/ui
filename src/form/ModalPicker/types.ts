export type AddButtonCallback<T> = () => Promise<T>;

export interface AddButtonOptions<T> {
  label: string;
  onClick: AddButtonCallback<T>;
}
