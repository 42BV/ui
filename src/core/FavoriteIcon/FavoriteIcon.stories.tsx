import { useState } from 'react';
import { storiesOf } from '@storybook/react';

import { FavoriteIcon } from './FavoriteIcon';

storiesOf('core/FavoriteIcon', module)
  .addParameters({ component: FavoriteIcon })
  .add('default', () => {
    const [favorite, setFavorite] = useState(false);
    return (
      <div className="text-center">
        <FavoriteIcon
          onChange={() => setFavorite(!favorite)}
          value={favorite}
        />
      </div>
    );
  })
  .add('hover color', () => {
    const [favorite, setFavorite] = useState(false);
    return (
      <div className="text-center">
        <FavoriteIcon
          onChange={() => setFavorite(!favorite)}
          value={favorite}
          hoverColor="danger"
        />
      </div>
    );
  });
