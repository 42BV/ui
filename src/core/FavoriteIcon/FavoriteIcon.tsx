import React from 'react';
import Icon from '../Icon/Icon';
import { useHover } from '../../hooks/useHover/useHover';
import { Color } from '../types';

type FavoriteIconProps = {
  /**
   * The value that the form element currently has.
   */
  value: boolean;

  /**
   * Callback for when the form element changes.
   */
  onChange: (value: boolean) => void;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;

  /**
   * Optionally the size of the icon in pixels.
   *
   * Defaults to `24px`
   */
  size?: number;

  /**
   * Optionally the color of the icon when value is false
   * and the icon is not hovered.
   *
   * Defaults to secondary.
   */
  color?: Color;

  /**
   * Optionally the color of the icon when value is true
   * or the icon is hovered.
   *
   * Defaults to primary.
   */
  activeColor?: Color;
};

/**
 * FavoriteIcon displays a star icon to show the user
 * whether an item is favorite or not. The component
 * includes toggle functionality.
 */
export function FavoriteIcon({
  value,
  onChange,
  className = '',
  size,
  color = 'secondary',
  activeColor = 'primary'
}: FavoriteIconProps) {
  const [hover, hoverEvents] = useHover();

  return (
    <div {...hoverEvents}>
      <Icon
        color={value || hover ? activeColor : color}
        className={className}
        icon={value ? 'star' : 'star_border'}
        onClick={() => onChange(!value)}
        size={size}
      />
    </div>
  );
}
