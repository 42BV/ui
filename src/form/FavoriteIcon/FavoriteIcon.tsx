import React from 'react';
import Icon from '../../core/Icon/Icon';
import { useHover } from '../../hooks/useHover/useHover';

type FavoriteIconProps = {
  /**
   * The value that the form element currently has.
   */
  value: boolean;

  /**
   * Callback for when the form element changes.
   */
  onChange: (value: boolean) => void;

  /**
   * Optional extra CSS class you want to add to the component.
   * Useful for styling the component.
   */
  className?: string;
};

/**
 * FavoriteIcon displays a star icon to show the user
 * whether an item is favorite or not. The component
 * includes toggle functionality.
 */
export function FavoriteIcon({
  value,
  onChange,
  className = ''
}: FavoriteIconProps) {
  const [hover, hoverEvents] = useHover();

  return (
    <div {...hoverEvents}>
      <Icon
        color={value || hover ? 'primary' : 'secondary'}
        className={className}
        icon={value ? 'star' : 'star_border'}
        onClick={() => onChange(!value)}
      />
    </div>
  );
}
