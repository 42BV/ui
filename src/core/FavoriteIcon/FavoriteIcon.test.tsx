import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { FavoriteIcon } from './FavoriteIcon';
import { Color } from '../types';

describe('Component: FavoriteIcon', () => {
  function setup({
    value,
    color,
    hoverColor,
    size
  }: {
    value: boolean;
    color?: Color;
    hoverColor?: Color;
    size?: number;
  }) {
    const onChangeSpy = jest.fn();

    const favoriteIcon = shallow(
      <FavoriteIcon
        onChange={onChangeSpy}
        value={value}
        color={color}
        hoverColor={hoverColor}
        size={size}
      />
    );

    return { favoriteIcon, onChangeSpy };
  }

  describe('ui', () => {
    test('favorite', () => {
      const { favoriteIcon } = setup({ value: true });

      expect(toJson(favoriteIcon)).toMatchSnapshot(
        'Component: FavoriteIcon => ui => favorite'
      );
    });

    test('not favorite', () => {
      const { favoriteIcon } = setup({ value: false });

      expect(toJson(favoriteIcon)).toMatchSnapshot(
        'Component: FavoriteIcon => ui => not favorite'
      );
    });

    test('size', () => {
      const { favoriteIcon } = setup({ value: false, size: 10 });

      expect(favoriteIcon.find('Icon').props().size).toBe(10);
    });

    test('color', () => {
      const { favoriteIcon } = setup({ value: false, color: 'primary' });

      expect(favoriteIcon.find('Icon').props().color).toBe('primary');
    });

    describe('hoverColor', () => {
      it('should use hoverColor when hoverColor is defined', () => {
        const { favoriteIcon } = setup({
          value: true,
          color: 'secondary',
          hoverColor: 'success'
        });

        // @ts-expect-error Mock test
        expect(favoriteIcon.find('Icon').props().hoverColor).toBe('success');
      });

      it('should use activeColor when hoverColor is not defined', () => {
        const { favoriteIcon } = setup({ value: true });

        // @ts-expect-error Mock test
        expect(favoriteIcon.find('Icon').props().hoverColor).toBe('primary');
      });
    });
  });

  describe('events', () => {
    it('should when value is false and the icon is clicked call onChange with true', () => {
      const { favoriteIcon, onChangeSpy } = setup({ value: false });

      const event = {
        preventDefault: jest.fn()
      };
      favoriteIcon.find('Icon').simulate('click', event);

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith(true);
    });

    it('should when value is true and the icon is clicked call onChange with false', () => {
      const { favoriteIcon, onChangeSpy } = setup({ value: true });

      const event = {
        preventDefault: jest.fn()
      };
      favoriteIcon.find('Icon').simulate('click', event);

      expect(onChangeSpy).toHaveBeenCalledTimes(1);
      expect(onChangeSpy).toHaveBeenCalledWith(false);
    });

    it('should prevent click callbacks in parent components from being called', () => {

      const { favoriteIcon } = setup({ value: true });

      const event = {
        preventDefault: jest.fn()
      };
      favoriteIcon.find('Icon').simulate('click', event);

      expect(event.preventDefault).toHaveBeenCalledTimes(1);
    });
  });
});
