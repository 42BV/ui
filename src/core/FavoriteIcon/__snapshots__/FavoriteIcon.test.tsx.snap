// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: FavoriteIcon ui favorite: Component: FavoriteIcon => ui => favorite 1`] = `
<Icon
  className=""
  color="primary"
  hoverColor="primary"
  icon="star"
  onClick={[Function]}
/>
`;

exports[`Component: FavoriteIcon ui not favorite: Component: FavoriteIcon => ui => not favorite 1`] = `
<Icon
  className=""
  color="secondary"
  hoverColor="primary"
  icon="star_border"
  onClick={[Function]}
/>
`;
